commit 85eab5a3f56c26fa89bde69c50ce55a08b345b9c
Author: Kalesh Singh <kaleshsingh@google.com>
Date:   Sun Oct 13 13:56:24 2024 -0700

    ltp: syscalls: mincore: Iterate vector in kernel page size granule
    
    Android emulates a userspace 16KiB page size on x86_64 emulators.
    The kernel however remains 4KiB and page faults are therefore handled
    in 4KiB granulatiry.
    
    Use the kernel_page_size() instead of the userspace page size when
    processing the mincore vector on Android's x86_64 16KiB page size
    emulators.
    
    This is effectively a no-op change on all other targets
    
    Bug: 366098040
    Test: atest vts_ltp_test_x86_64:syscalls.mincore03_64bit#syscalls.mincore03_64bit
    Test: atest vts_ltp_test_x86_64:syscalls.mincore04_64bit#syscalls.mincore04_64bit
    Change-Id: I03596bdef8b5e8e8141b51b1f5256eed04d1b368
    Signed-off-by: Kalesh Singh <kaleshsingh@google.com>

diff --git a/testcases/kernel/syscalls/mincore/mincore03.c b/testcases/kernel/syscalls/mincore/mincore03.c
index 774fce98b..c3ef2d203 100644
--- a/testcases/kernel/syscalls/mincore/mincore03.c
+++ b/testcases/kernel/syscalls/mincore/mincore03.c
@@ -16,6 +16,7 @@
 #include <unistd.h>
 #include <sys/mman.h>
 #include "tst_test.h"
+#include "pgsize_helpers.h"
 
 #define NUM_PAGES 3
 
@@ -46,7 +47,7 @@ static void setup(void)
 static void test_mincore(unsigned int test_nr)
 {
 	const struct tcase *tc = &tcases[test_nr];
-	unsigned char vec[NUM_PAGES];
+	DECLARE_MINCORE_VECTOR(vec, NUM_PAGES);
 	int locked_pages;
 	int count, mincore_ret;
 
@@ -58,9 +59,10 @@ static void test_mincore(unsigned int test_nr)
 	if (mincore_ret == -1)
 		tst_brk(TBROK | TERRNO, "mincore failed");
 	locked_pages = 0;
-	for (count = 0; count < NUM_PAGES; count++)
+	for (count = 0; count < nr_pgs_to_nr_kernel_pgs(NUM_PAGES); count++)
 		if (vec[count] & 1)
 			locked_pages++;
+	locked_pages = nr_kernel_pgs_to_nr_pgs(locked_pages);
 
 	if (locked_pages == tc->expected_pages)
 		tst_res(TPASS, "mincore() reports %s", tc->desc);
diff --git a/testcases/kernel/syscalls/mincore/mincore04.c b/testcases/kernel/syscalls/mincore/mincore04.c
index ed0ab7dfa..f718301fb 100644
--- a/testcases/kernel/syscalls/mincore/mincore04.c
+++ b/testcases/kernel/syscalls/mincore/mincore04.c
@@ -20,6 +20,7 @@
 #include <fcntl.h>
 #include <stdlib.h>
 #include "tst_test.h"
+#include "pgsize_helpers.h"
 
 #define NUM_PAGES 3
 
@@ -64,17 +65,18 @@ static int count_pages_in_cache(void)
 {
 	int locked_pages = 0;
 	int count, ret;
-	unsigned char vec[NUM_PAGES];
+	DECLARE_MINCORE_VECTOR(vec, NUM_PAGES);
 
 	TST_CHECKPOINT_WAIT(0);
 
 	ret = mincore(ptr, size, vec);
 	if (ret == -1)
 		tst_brk(TBROK | TERRNO, "mincore failed");
-	for (count = 0; count < NUM_PAGES; count++) {
+	for (count = 0; count < nr_pgs_to_nr_kernel_pgs(NUM_PAGES); count++) {
 		if (vec[count] & 1)
 			locked_pages++;
 	}
+	locked_pages = nr_kernel_pgs_to_nr_pgs(locked_pages);
 
 	TST_CHECKPOINT_WAKE(1);
 	return locked_pages;
